pyartcd/pyartcd/pipelines/gen_assembly.py

```diff
@@ -0,0 +1,233 @@
+from pyartcd.cli import cli, click_coroutine, pass_runtime
+from pyartcd.git import GitRepository
+from pyartcd.runtime import Runtime
+from ruamel.yaml import YAML
+
+yaml = YAML(typ="rt")
+
+
+class GenAssemblyPipeline:
+    """ Rebase and build MicroShift for an assembly """
+
+    def __init__(self, runtime: Runtime, group: str, assembly: str, ocp_build_data_url: str,
+                 nightlies: Tuple[str, ...], allow_pending: bool, allow_rejected: bool, allow_inconsistency: bool,
+                 custom: bool, arches: Tuple[str, ...], in_flight: Optional[str], previous_list: Tuple[str, ...], auto_previous: bool,
+                 logger: Optional[logging.Logger] = None):
+        self.runtime = runtime
+        self.group = group
+        self.assembly = assembly
+        self.nightlies = nightlies
+        self.allow_pending = allow_pending
+        self.allow_rejected = allow_rejected
+        self.allow_inconsistency = allow_inconsistency
+        self.custom = custom
+        self.arches = arches
+        self.in_flight = in_flight
+        self.previous_list = previous_list
+        self.auto_previous = auto_previous
+        self._logger = logger or runtime.logger
+        self._slack_client = self.runtime.new_slack_client()
+        self._working_dir = self.runtime.working_dir.absolute()
+
+        # determines OCP version
+        match = re.fullmatch(r"openshift-(\d+).(\d+)", group)
+        if not match:
+            raise ValueError(f"Invalid group name: {group}")
+        self._ocp_version = (int(match[1]), int(match[2]))
+
+        # sets environment variables for Doozer
+        self._doozer_env_vars = os.environ.copy()
+        self._doozer_env_vars["DOOZER_WORKING_DIR"] = str(self._working_dir / "doozer-working")
+
+        if not ocp_build_data_url:
+            ocp_build_data_url = self.runtime.config.get("build_config", {}).get("ocp_build_data_url")
+        if ocp_build_data_url:
+            self._doozer_env_vars["DOOZER_DATA_PATH"] = ocp_build_data_url
+
+    async def run(self):
+        self._slack_client.bind_channel(self.group)
+        slack_response = await self._slack_client.say(f":construction: Generating assembly definition {self.assembly} :construction:")
+        slack_thread = slack_response["message"]["ts"]
+        try:
+            if self.arches and self.custom:
+                raise ValueError("Customizing arches can only be used with custom assemblies.")
+
+            self._logger.info("Getting nightlies from Release Controllers...")
+            candidate_nightlies = await self._get_nightlies()
+            self._logger.info("Generating assembly definition...")
+            assembly_definition = await self._gen_assembly_from_releases(candidate_nightlies)
+            out = StringIO()
+            yaml.dump(assembly_definition, out)
+            self._logger.info("Generated assembly definition:\n%s", out.getvalue())
+
+            # Create a PR
+            def _merge(a, b):
```

Comment:
Looks like a utility function we used to merge two Python dictionaries. A docstring (even a concise one) would be appreciated, since recursion could make it not straight forward to understand what is going on

Reply:
I moved it out of the member function.